### ISSUE TITLE
An error should be raised for bad implementations of `Display` in an unit struct context

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust#68968]
This snippet does show an error:

```rust
use core::fmt;

struct Foo(i32);

impl fmt::Display for Foo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self)
    }
}

fn main() {
    println!("{}", Foo);
}
```

But the following overflows the stack:

```rust
use core::fmt;

struct Foo;

impl fmt::Display for Foo {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self)
    }
}

fn main() {
    println!("{}", Foo);
}
```

**Comment by jonas-schievink** [comment]
The first one shows

    error[E0277]: `fn(i32) -> Foo {Foo}` doesn't implement `std::fmt::Display`

because you're passing the type constructor function, not an instance of `Foo`. Change it to `Foo(123)` and it also overflows the stack.

The unconditional recursion lint not triggering on `Display` impls is a well known issue tracked (primarily) by https://github.com/rust-lang/rust/issues/57965, so closing in favor of that.
